### ISSUE TITLE
LoRA Training

### ISSUE THREAD

**Issue by Galarzaa90** [Knowledgator/GLiClass#28]
Hello! Thanks for releasing GLiClass.

I’m fine-tuning GLiClass with LoRA and started from [train.py](https://github.com/Knowledgator/GLiClass/blob/main/train.py). Do you have an official LoRA/QLoRA script or recommended configuration? If not, any quick pointers on:

- Target modules to adapt
- Typical LoRA parameters
- Precision/memory tips 

Thanks!

PS: Is this the right channel for this? Please let me know if there's a better place for these kind of inquiries.

**Comment by BioMikeUkr** [comment]
Hello, you could start with the parameters we used:
|               | [gliclass‑modern‑base‑v3.0](https://huggingface.co/knowledgator/gliclass-modern-base-v3.0) | [gliclass‑modern‑large‑v3.0](https://huggingface.co/knowledgator/gliclass-modern-large-v3.0) | [gliclass‑base‑v3.0](https://huggingface.co/knowledgator/gliclass-base-v3.0) | [gliclass‑large‑v3.0](https://huggingface.co/knowledgator/gliclass-large-v3.0) |
|----------------------|---------------------------------|----------------------------------|--------------------------------|---------------------------------|
| LoRa r               | 512 | 768 | 384 | 384 |
| LoRa α               | 1024 | 1536 | 768 | 768 |
| focal loss α         | 0.7 | 0.7 | 0.7 | 0.7 |
| Target modules       | "Wqkv", "Wo", "Wi", "linear_1", "linear_2" |  "Wqkv", "Wo", "Wi", "linear_1", "linear_2"  |  "query_proj", "key_proj", "value_proj", "dense", "linear_1", "linear_2", mlp.0", "mlp.2", "mlp.4" |  "query_proj", "key_proj", "value_proj", "dense", "linear_1", "linear_2", mlp.0", "mlp.2", "mlp.4" |

The key idea is that the larger the LoRa r, the fewer data and epochs you need to fine-tune the model. Try LoRa a between 256-1024. Probably we'll prepare a script for fine-tuning with Peft.

**Comment by BioMikeUkr** [comment]
> a finetuning script with peft would be really helpful, could you also mention how much time it token for finetuning with the gpu and LoRA of such high ranks? if the ranks are so high why not use the traditional finetuning approach?

Here you can find some useful calculations for `knowledgator/gliclass-edge-v3.0` for different LoRa a (32-2048, r=0 - without LoRa). Fine-tuning with LoRa could be 30-40% more time-efficient (medium-large models). However, the main reason is that there is a crossover, where training time with LoRA is lower than without it, and eval accuracy is higher than without LoRA. With LoRa, models are fine-tuned more stably.

<img width="1280" height="819" alt="Image" src="https://github.com/user-attachments/assets/b7950e3c-6a21-4c2a-80e3-bcb24e635534" />

<img width="1280" height="818" alt="Image" src="https://github.com/user-attachments/assets/97cb2390-5d25-4ec6-b27e-8e87c8334153" />

<img width="1280" height="820" alt="Image" src="https://github.com/user-attachments/assets/ee8c40a6-6181-42c4-aa7b-13d1fa9e2ec3" />
